### PR TITLE
Adds new resource cloudflare_healthcheck

### PR DESCRIPTION
**Pull request by cehrig**
This PR introduces a new resource `cloudflare_standalone_healthcheck` for creating Standalone Health Checks without needing a Cloudflare Load Balancer.

An HTTPs and TCP example is provided in the website-doc. @jacobbednarz @patryk this PR is rather large, but I'd appreciate if you could already review the resource and have a look if that goes into the right direction.

Todo:
- Write acceptance test

#645

### CODE REVIEW

**Comment by jacobbednarz** [comment]
Sure thing! I'll try to get to this today.

File: cloudflare/provider.go [hunk 1]
```diff
@@ -126,6 +126,7 @@ func Provider() terraform.ResourceProvider {
 			"cloudflare_rate_limit":                  resourceCloudflareRateLimit(),
 			"cloudflare_record":                      resourceCloudflareRecord(),
 			"cloudflare_spectrum_application":        resourceCloudflareSpectrumApplication(),
+			"cloudflare_standalone_healthcheck":      resourceCloudflareStandaloneHealthcheck(),
```
Comment (jacobbednarz) [inline cloudflare/provider.go:129]: mayyybe?

```suggestion
			"cloudflare_healthcheck":      resourceCloudflareStandaloneHealthcheck(),
```

File: cloudflare/resource_cloudflare_standalone_healthcheck.go [hunk 2]
```diff
@@ -0,0 +1,396 @@
...
+		Schema: map[string]*schema.Schema{
+			"zone_id": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
```
Comment (jacobbednarz) [inline cloudflare/resource_cloudflare_standalone_healthcheck.go:34]: Probably doesn't matter if the name changes that much

```suggestion
```

File: cloudflare/provider.go (hunk 1 above)
Comment (cehrig) [reply cloudflare/provider.go]: Absolutely.

File: cloudflare/resource_cloudflare_standalone_healthcheck.go (hunk 2 above)
Comment (cehrig) [reply cloudflare/resource_cloudflare_standalone_healthcheck.go]: Indeed. I was under the false impression we can't update the name.

**Comment by cehrig** [comment]
I've added some acceptance tests inspired by the loadbalancer monitor test
- Create simple TCP Healthcheck
- Update TCP Healthcheck
- Create simple HTTP healthcheck
- Trigger error when no name provided

Happy to add more

File: cloudflare/resource_cloudflare_healthcheck_test.go [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+	rnd := generateRandomResourceName()
+	name := fmt.Sprintf("cloudflare_healthcheck.%s", rnd)
+	var healthcheck cloudflare.Healthcheck
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:  func() { testAccPreCheck(t) },
+		Providers: testAccProviders,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccCheckCloudflareHealthcheckTCP(zoneID, "tcp-test-healthcheck", rnd),
```
Comment (jacobbednarz) [inline cloudflare/resource_cloudflare_healthcheck_test.go:27]: We can sub out `"tcp-test-healthcheck"` with `rnd` here to ensure the name is always unique. In the cases of updated, I generally do something like `rnd + "-updated"` to be able to track the differences if that is something your tests need.

File: cloudflare/resource_cloudflare_healthcheck_test.go [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+	name := fmt.Sprintf("cloudflare_healthcheck.%s", rnd)
+	var healthcheck cloudflare.Healthcheck
+	var initialID string
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:  func() { testAccPreCheck(t) },
+		Providers: testAccProviders,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccCheckCloudflareHealthcheckTCP(zoneID, "tcp-test-healthcheck-update", rnd),
```
Comment (jacobbednarz) [inline cloudflare/resource_cloudflare_healthcheck_test.go:53]: This is a good example of the second part I mentioned in https://github.com/terraform-providers/terraform-provider-cloudflare/pull/680/files#r423320742. We could use it here to have a guaranteed unique name and then confirm it in the config by appending `-updated` where needed.

File: cloudflare/resource_cloudflare_healthcheck_test.go [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+	rnd := generateRandomResourceName()
+	name := fmt.Sprintf("cloudflare_healthcheck.%s", rnd)
+	var healthcheck cloudflare.Healthcheck
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:  func() { testAccPreCheck(t) },
+		Providers: testAccProviders,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccCheckCloudflareHealthcheckHTTP(zoneID, "http-test-healthcheck", rnd),
```
Comment (jacobbednarz) [inline cloudflare/resource_cloudflare_healthcheck_test.go:93]: Same comment as https://github.com/terraform-providers/terraform-provider-cloudflare/pull/680/files#r423320742

File: website/docs/r/healthcheck.html.markdown [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+# cloudflare_healthcheck
+
+Standalone Health Checks provide a way to monitor origin servers without needing a Cloudflare Load Balancer. 
+
+## Example Usage
+
+The resource supports HTTP, HTTPS and TCP type health checks.
+
+### HTTPS Health Check
+```hcl
```
Comment (jacobbednarz) [inline website/docs/r/healthcheck.html.markdown:18]:
```suggestion

```hcl
```

File: website/docs/r/healthcheck.html.markdown [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+  timeout = 10
+  retries = 2
+  interval = 60
+  consecutive_fails = 3
+  consecutive_successes = 2
+}
+```
+
+### TCP Monitor
+```hcl
```
Comment (jacobbednarz) [inline website/docs/r/healthcheck.html.markdown:57]:
```suggestion

```hcl
```

File: website/docs/r/healthcheck.html.markdown [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+* `type` - (Required) The protocol to use for the health check. Valid values: `HTTP`, `HTTPS`, `TCP`.
+* `port` - (Optional) Port number to connect to for the health check.  Valid values are in the rage `0-65535` (Default: `80`).
+* `timeout` - (Optional) The timeout (in seconds) before marking the health check as failed. (Default: `5`)
+* `retries` - (Optional) The number of retries to attempt in case of a timeout before marking the origin as unhealthy. Retries are attempted immediately. (Default: `2`)
+* `interval` - (Optional) The interval between each health check. Shorter intervals may give quicker notifications if the origin status changes, but will increase load on the origin as we check from multiple locations. (Default: `60`)
+* `consecutive_fails` - (Optional) The number of consecutive fails required from a health check before changing the health to unhealthy. (Default: `1`)
+* `consecutive_successes` - (Optional) The number of consecutive successes required from a health check before changing the health to healthy. (Default: `1`)
+
+### HTTP/HTTPS specific arguments
```
Comment (jacobbednarz) [inline website/docs/r/healthcheck.html.markdown:103]:
```suggestion

### HTTP/HTTPS specific arguments
```

File: website/docs/r/healthcheck.html.markdown [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+* `follow_redirects` - (Optional) Follow redirects if the origin returns a 3xx status code. Valid values: `true` or `false` (Default: `false`).
+* `allow_insecure` - (Optional) Do not validate the certificate when the health check uses HTTPS. Valid values: `true` or `false` (Default: `false`).
+* `header` - (Optional) The HTTP request headers to send in the health check. It is recommended you set a Host header by default. The User-Agent header cannot be overridden.
+
+**header** requires the following:
+
+* `header` - (Required) The header name.
+* `values` - (Required) A list of string values for the header.
+
+### TCP specific arguments
```
Comment (jacobbednarz) [inline website/docs/r/healthcheck.html.markdown:117]:
```suggestion

### TCP specific arguments
```

File: cloudflare/resource_cloudflare_healthcheck_test.go (hunk 3 above)
Comment (cehrig) [reply cloudflare/resource_cloudflare_healthcheck_test.go]: Updated accordingly.

File: cloudflare/resource_cloudflare_healthcheck_test.go (hunk 4 above)
Comment (cehrig) [reply cloudflare/resource_cloudflare_healthcheck_test.go]: Updated here as well

File: cloudflare/resource_cloudflare_healthcheck_test.go (hunk 5 above)
Comment (cehrig) [reply cloudflare/resource_cloudflare_healthcheck_test.go]: Updated here as well

**Comment by cehrig** [comment]
Yeah, I was on the edge of creating nested structs but duplication of attributes + line up with the loadbalancer monitors made me follow this approach.

I've updated the tests with your remarks and merged your doc suggestions + completely overlooked that it's incompatible with API tokens, thanks for that fix!

**Comment by jacobbednarz** [comment]
Thanks a bunch! This looks awesome 🤜🤛

**Pull request merged**